htlcswitch/switch.go

```diff
@@ -367,6 +367,14 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID,
 		return zeroPreimage, err
 	}
 
+	// If we already know the preimage, we can return it immediately.
+	p, ok := s.cfg.PreimageCache.LookupPreimage(htlc.PaymentHash[:])
+	if ok {
+		var preImg [32]byte
+		copy(preImg[:], p[:])
+		return preImg, nil
```

Comment:
How is this going to work when we not only need to get the preimage from the previous successful payment attempt, but also the route taken?

Reply:
Okay, you explained it nicely above. It is stored persistently by the router.